### PR TITLE
Adapt text generation fct. to new tokenizer API

### PR DESCRIPTION
**Pull request by mali-git**

### CODE REVIEW

**Comment by le1nux** [comment]
* separated checkpointing into checkpoint loading and checkpoint saving. The reason is the for checkpoint loading at inference time, we don't need the checkpoint saving functionality. 
* introduced a inference component for the text modularity. Others modularities can be added. At the moment the model is loaded on a single GPU. Later on, we can support more sophisticated serving environments
* adapted the text generation interface to work with a YAML config that specifies the tokeniser, model etc.

File: src/modalities/checkpointing/torch/torch_checkpoint_loading.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+class TorchCheckpointLoading(CheckpointLoadingIF):
+    def __init__(self, device: torch.device, precision: PrecisionEnum):
+        self.device = device
+        self.precision = precision
+
+    def load_model_checkpoint(self, model: nn.Module, file_path: Path) -> nn.Module:
+        model_state = torch.load(file_path)
+        model.load_state_dict(model_state)
+        # set the model to the correct device and precision
+        model = model.to(self.precision.value)
```
Comment (le1nux) [inline src/modalities/checkpointing/torch/torch_checkpoint_loading.py:21]: check if we have to set the precision explicitly

File: src/modalities/config/config.py [hunk 2]
```diff
@@ -115,7 +125,39 @@ class SaveKMostRecentCheckpointsStrategyConfig(BaseModel):
     k: Annotated[int, Field(strict=True, ge=-1)]
 
 
-class FSDPToDiscCheckpointingConfig(BaseModel):
+class TorchCheckpointLoadingConfig(BaseModel):
+    device: PydanticPytorchDeviceType
+    precision: PrecisionEnum
```
Comment (le1nux) [inline src/modalities/config/config.py:130]: check if we need precision

File: src/modalities/inference/text/inference_component.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        self.temperature = temperature
+        self.context_length = context_length
+
+    def generate_tokens(
+        self,
+        context: str,
+    ):
+        token_ids_list = self.tokenizer.tokenize(context)
+        max_new_tokens = self.context_length - len(token_ids_list)
+        input_token_ids = torch.IntTensor(token_ids_list).cuda().unsqueeze(0)
```
Comment (le1nux) [inline src/modalities/inference/text/inference_component.py:34]: coda might not be applicable in all cases

File: src/modalities/inference/text/inference_component.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        self.eod_token = eod_token
+        self.prompt_template = prompt_template
+        self.temperature = temperature
+        self.context_length = context_length
+
+    def generate_tokens(
+        self,
+        context: str,
+    ):
+        token_ids_list = self.tokenizer.tokenize(context)
```
Comment (le1nux) [inline src/modalities/inference/text/inference_component.py:32]: check what happens if we pass in an empty context

File: src/modalities/inference/text/inference_component.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        # Dictionary to hold variable names and user provided values
+        user_inputs = {}
+
+        # Ask user for input for each found variable name
+        for var in variable_names:
+            user_inputs[var] = input(f"{var}: ")
+
+        # Use str.format() to replace placeholders with user values
+        formatted_string = template.format(**user_inputs)
+        return formatted_string
```
Comment (le1nux) [inline src/modalities/inference/text/inference_component.py:93]: test that formatted_string is not empty

File: src/modalities/inference/text/inference_component.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            # idx_next = torch.multinomial(probs, num_samples=1)
+            idx_next = torch.argmax(logits, dim=-1)
+            # token_id: int = idx_next[0, 0].item()
+            token_id: int = idx_next.item()
+            generated_token_ids.append(token_id)
+            idx_next_str = self.tokenizer.decode([token_id])
+            generated_text_new = self.tokenizer.decode(generated_token_ids)
+
+            if idx_next_str == self.eod_token:
+                print("\n<reached eos token>", end="")
```
Comment (le1nux) [inline src/modalities/inference/text/inference_component.py:57]: end of document

File: tests/checkpointing/torch/test_torch_checkpoint_loading.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+def test_load_model_checkpoint(tmp_path):
+    tmp_file_path = tmp_path / "model_state.pth"
+
+    model = DummyModel()
+
+    model_state = model.state_dict()
+    torch.save(model_state, tmp_file_path)
+
+    loaded_model: DummyModel = TorchCheckpointLoading(
+        device=torch.device("cpu"), precision=PrecisionEnum.FP16
```
Comment (le1nux) [inline tests/checkpointing/torch/test_torch_checkpoint_loading.py:29]: add precision tests to the dummy model

File: src/modalities/config/utils.py [hunk 8]
```diff
@@ -1,8 +1,18 @@
...
     return {key: getattr(config, key) for key in config.model_dump().keys()}
+
+
+def parse_torch_device(device: str | int) -> torch.device:
+    if isinstance(device, str) and device != "cpu":
+        raise ValueError(f"Invalid device_id: {device}")
+    else:
+        device_id = f"cuda:{device}"
+    device = torch.device(device_id)
+    return device
```
Comment (flxst) [inline src/modalities/config/utils.py:18]: This doesn't seem right. What if `device == "cpu"`?

File: data/sample_datasets/redpajama_v2/mem_map/redpajama_v2_gpt2_tokenized_num_samples_512_test.idx
Comment (flxst) [inline data/sample_datasets/redpajama_v2/mem_map/redpajama_v2_gpt2_tokenized_num_samples_512_test.idx:1]: What is the purpose of adding this and the 3 other *.idx / *.pbin files? I can't see them being used anywhere.

File: src/modalities/config/utils.py (hunk 8 above)
Comment (le1nux) [reply src/modalities/config/utils.py:18]: true, fixed it.

File: data/sample_datasets/redpajama_v2/mem_map/redpajama_v2_gpt2_tokenized_num_samples_512_test.idx
Comment (le1nux) [reply data/sample_datasets/redpajama_v2/mem_map/redpajama_v2_gpt2_tokenized_num_samples_512_test.idx:1]: it was removed :)

**Pull request merged**